_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,421 @@
+- All storage slots are set to zero at the beginning of the tests (meanings tests are effectively run against an empty blockchain)
+
+In fact, the only symbolic variables introduced into the test environment are those that are
+specified in the signature of the test method. This means that the proofs are exhaustive *only over
+the input variables*. As an example, consider the `prove_transfer` test from the example above. The
+`totalSupply` is always `uint(-1)`, and the test would not catch obviously faulty implementations of
+`transfer` like the one below:
+
+```solidity
+function transfer(address dst, uint amt) public {
+    require(totalSupply > uint56(-1), "whoops");
+    balanceOf[msg.sender] = sub(balanceOf[msg.sender], amt);
+    balanceOf[dst]        = add(balanceOf[dst], amt);
+}
+```
+
+## Execution Against Mainnet State
+
+`hevm` allows us to fetch state from an rpc node, and we can also use this to write symbolic tests
+against mainnet state. As an example, lets run `prove_transfer` against the [balancer
+token](https://etherscan.io/address/0xba100000625a3754423978a60c9317c58a424e3D#code):
+
+```solidity
+interface ERC20 {
+    function totalSupply() external view returns (uint256);
+    function balanceOf(address account) external view returns (uint256);
+    function transfer(address recipient, uint256 amount) external returns (bool);
+    function allowance(address owner, address spender) external view returns (uint256);
+    function approve(address spender, uint256 amount) external returns (bool);
+    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);
+    event Transfer(address indexed from, address indexed to, uint256 value);
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+}
+
+contract TestToken is DSTest, SafeMath {
+    function prove_transfer(address dst, uint amt) public {
+        // BAL: https://etherscan.io/address/0xba100000625a3754423978a60c9317c58a424e3D#code
+        ERC20 token = ERC20(0xba100000625a3754423978a60c9317c58a424e3D);
+
+        uint preBalThis = token.balanceOf(address(this));
+        uint preBalDst  = token.balanceOf(dst);
+
+        token.transfer(dst, amt);
+
+        // no change for self-transfer
+        uint delta = dst == address(this) ? 0 : amt;
+
+        // balance of this has been reduced by `delta`
+        assertEq(sub(preBalThis, token.balanceOf(address(this))), delta);
+
+        // balance of dst has been increased by `delta`
+        assertEq(sub(token.balanceOf(dst), preBalDst), delta);
+    }
+
+}
+```
+
+We can execute this test by running `dapp test --rpc-url <URL>`, and we see that a counter example
+is found: the balancer token reverts for transfers to the zero address.
+
+```
+[FAIL] prove_transfer(address,uint256)
+
+Failure: prove_transfer(address,uint256)
+
+  Counter Example:
+
+    result:   Revert
+    calldata: prove_transfer(0x0000000000000000000000000000000000000000, 0)
+```
+
```

Comment:
Nice. This is a cool example. link to https://github.com/xwvvvvwx/weird-erc20? :)